tool/tsh/common/db.go

```diff
@@ -1248,115 +1292,95 @@ func isMFADatabaseAccessRequired(ctx context.Context, tc *client.TeleportClient,
 //
 // If logged into multiple databases, returns an error unless one specified
 // explicitly via --db flag.
-func pickActiveDatabase(cf *CLIConf, tc *client.TeleportClient) (*tlsca.RouteToDatabase, error) {
-	profile, err := tc.ProfileStatus()
-	if err != nil {
-		return nil, trace.Wrap(err)
+func pickActiveDatabase(cf *CLIConf, tc *client.TeleportClient, routes []tlsca.RouteToDatabase) (*tlsca.RouteToDatabase, error) {
+	selectors := newDatabaseResourceSelectors(cf)
+	// first check if an active route can even potentially match.
+	routes = filterRoutesByPrefix(routes, selectors.name)
+	if len(routes) == 0 {
+		return nil, trace.NotFound(formatDbNotLoggedIn(cf.SiteName, selectors))
 	}
-
-	routes, err := profile.DatabasesForCluster(tc.SiteName)
+	info, err := getDatabaseInfo(cf, tc, routes)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	if !info.isActive {
+		return nil, trace.NotFound(formatDbNotLoggedIn(cf.SiteName, selectors))
+	}
```

Comment:
The input parameter for this function is `routes []tlsca.RouteToDatabase`   that orginates from 
```
routes, err := profile.DatabasesForCluster(tc.SiteName) 
``` 
notice that the `DatabasesForCluster`  call already returns active databases   [findActiveDatabases](https://github.com/gravitational/teleport/blob/master/lib/client/profile.go#L554)

Why we need to check isActive one more time ?

Reply:
Because it calls `getDatabaseInfo` - that func will resolve the name, labels, and query selectors to a single database and then search for that database's full name in the active db routes passed to it. `getDatabaseInfo` doesn't necessarily find an active db though, it just looks for one so it can copy the db user/schema from an active route

Reply:
What happens in this case?

0. user has access to `mysql` and mysql2`?
1. `tsh db login mysql --db-user user`
2. `tsh db login mysql2 --db-user user`
3. remove `mysql` from "backend"
4. `tsh db config mysql`

Or the same sequence without #2. 

I thought I was somewhat familiar with the recent changes, but I am having a hard time tracking cases through these calls =p.

Reply:
If you `tctl rm db/mysql` in step 2, then in step 4 you will get an error since you are not logged into `foo2` which matches by prefix now:
```
$ tsh db ls env=test 
Name Description                   Allowed Users Labels   Connect 
---- ----------------------------- ------------- -------- ------- 
foo  Example dyn database resource [*]           env=test         
foo2 Example dyn database resource [*]           env=test         

$ tsh db login --db-user=alice foo 
Connection information for database "foo" has been saved.

You can now connect to it using the following command:

  tsh db connect foo

You can view the connect command for the native database CLI client:

  tsh db config --format=cmd foo

$ tctl rm db/foo 
database "foo" has been deleted

$ tsh db config foo 
ERROR: please provide the database user using the --db-user flag, allowed database users for foo2: [*]
```

If you don't delete the backend database, then it works:
```
$ tsh db config foo 
Name:      foo
Host:      gavin-leaf.cloud.gravitational.io
Port:      3080
User:      alice
Database:  
CA:        /Users/gavin/.tsh/keys/gavin-leaf.cloud.gravitational.io/cas/gavin-leaf.cloud.gravitational.io.pem
Cert:      /Users/gavin/.tsh/keys/gavin-leaf.cloud.gravitational.io/gavin.frazar@goteleport.com-db/gavin-leaf.cloud.gravitational.io/foo-x509.pem
Key:       /Users/gavin/.tsh/keys/gavin-leaf.cloud.gravitational.io/gavin.frazar@goteleport.com
```

Reply:
> edit: oh, you asked about logging into both. In that case, it will display config for mysql2

I just ran into a similar situation

1.  `tsh db connect self-hosted-redis`
2.  `self-hosted-redis` from my config, restart db agent
3.  `tsh db connect self-hosted-redis` connects to `self-hosted-redis-cluster`.

I would rather see:
```
$ tsh db connect self-hosted-redis
ERROR: database "self-hosted-redis" not found, use 'tsh db ls' to see registered databases
```

> I figured this might screw up a script after the v13->v14 upgrade.

Personally I am not that worried about stale certs. I am sure it happens today even without the renaming change when admin removes a database. I think the above not-found error is very clear.

Here is my recommendation to adjust the logic for `tsh db connect/config/env/logout`/`tsh proxy db`

- If `--labels/--query` is specified for, the API must return exactly one DB (same as now?)
- If `database-service-name` is specified
  - Prefer exact full name match
  - Then exactly one discovered name match
  - Do NOT allow any other prefix matching. like input `mysq` shouldn't connect a database `mysql`.
- `database-service-name` vs `--labels/--query` should be mutually exclusive

@smallinsky @GavinFrazar what do you guys think?

Reply:
> I just ran into a similar situation

> tsh db connect self-hosted-redis
> self-hosted-redis from my config, restart db agent
> tsh db connect self-hosted-redis connects to self-hosted-redis-cluster.

> Here is my recommendation to adjust the logic for tsh db connect/config/env/logout/tsh proxy db

> If --labels/--query is specified, the API must return exactly one DB (same as now?)
If database-service-name is specified
Prefer exact full name match
Then exactly one discovered name match
Do NOT allow any other prefix matching. like input mysq shouldn't connect a database mysql.
database-service-name vs --labels/--query should be mutually exclusive

This can also also happen for `database-service-name` for instance if  the `rds-instance01` db has been  removed/deleted and another db was added with the shared db name prefix like `rds-instance02` the ` tsh db connect rds-instance` will connect to `rds-instance02` 

But for backward compatibility this is safer because we will not silently select  unintended db. So I agree that for backward capability this logic is better.

Reply:
> database-service-name vs --labels/--query should be mutually exclusive
Why?

Reply:
> Why ?

👍 hmm. or use the stricter rules when both are specified. I guess `tsh db connect rds --labels env=prod` is a very good use case

Reply:
@smallinsky @greedy52  I reworked the selection logic to follow Steve's suggestions here: faed6ca 

- match full name exactly
- OR match an unambiguous "discovered name"
- no prefix matching.

This preserves backwards compatibility, without being a nightmare of edge cases and unexpected or suprising UX.
Great suggestion Steve!

If these changes are approved then I'll do the same for `tsh kube` and `tctl` as well, which should be a much smaller and easier change for those commands.

Reply:
one caveat that I discussed with Steve over slack:

if a user has to resolve an ambiguous discovered name with a label to login, the logged in db can resolve ambiguity with further commands.

For example:

```
suppose you have databases like
...
foo-rds-us-west-1 (short name: foo)
foo-rds-us-west-2 (short name: foo)

$ tsh db login foo
...<ambiguous db error>...

$ tsh db login foo --labels region=us-west-1
...<logs into foo-rds-us-west-1>...

$ tsh db connect foo
...connects to foo-rds-us-west-1...
```

This seems more consistent UX, given that we support `tsh db connect` auto-selecting an active route.